### PR TITLE
feat(sdk): support track sdk for experiment

### PR DESCRIPTION
**Pull request by tianweidut**
## Description
- basic track SDK code skeleton:
   - only supports five core functions: `track.start, track.end, track.metrics, track.artifacts, track.params`.
   - only store results in local storage.
- mnist example:
  - [asciicast](https://asciinema.org/a/555472)

## Modules
- [x] Client
- [x] Python-SDK

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/api/_impl/track/collector.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            if not v:
+                continue
+
+            data[f"{k}/last"] = float(v[-1])
+            data[f"{k}/avg"] = round(sum(v) / len(v), 3)
+
+        if not data:
+            return
+
+        self.tracker._log_metrics(
```
Comment (goldenxinxing) [inline client/starwhale/api/_impl/track/collector.py:230]: should the func named to 'log_metrics'?

File: client/starwhale/api/_impl/track/tracker.py [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+
+        _step: int = self._metric_step
+        if step is not None:
+            if not isinstance(step, int) or step < 0:
+                raise TypeError(
+                    f"step({step}) is unexpected, the normal step is non-negative integer"
+                )
+            _step = step
+
+        with self._metric_lock:
```
Comment (goldenxinxing) [inline client/starwhale/api/_impl/track/tracker.py:335]: Is it not necessary to use this lock if we use _data directly?

File: client/starwhale/api/_impl/track/base.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        self,
+        name: str,
+        index: int,
+        created_at: str,
+        data: Link,
+        link_wrapper: bool = True,
+    ) -> None:
+        self.name = name
+        self.index = index
+        self.key = f"{name}-{index:03}"
```
[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/api/_impl/track/base.py:107]:
```
In [13]: [f"{10**i:03}" for i in range(0, 5)]
Out[13]: ['001', '010', '100', '1000', '10000']
```
1. when the value is more than 999, it sill show the origin value.
2. The refill zero operation is friendly for the datastore name sorting. 
3. In one track, the total number of versions of one artifact was less than 1000 in common case.

File: client/starwhale/api/_impl/track/tracker.py [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+    @classmethod
+    def start(
+        cls,
+        name: str = "",
+        project_uri: t.Optional[t.Union[str, URI]] = None,
+        access_token: str = "",
+        auto_run_collect: bool = True,
+        auto_framework_hook: bool = True,
+        auto_console_log_redirect: bool = True,
+        mode: str = "online",
```
[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/api/_impl/track/tracker.py:154]: "offline" mode only stores the original data in the local dir which needs be synced by the users.

- "online" -> the data in the Standalone instance or Cloud instance, for showing
- "offline" -> the data in the local '.starwhale 'dir, for storing.

When the user config "offline" mode, the syncer will not work. The Standalone instance is not equal to “offline mode”. In the typical case, the syncer is necessary. The other experiment SDKs also support the same feature and the default value.

File: client/starwhale/api/_impl/track/collector.py (hunk 1 above)
Comment (tianweidut) [reply client/starwhale/api/_impl/track/collector.py:230]: `_log_metrics` is an internal function. The user should use `track.metrics()` function.

`track._log_metrics` -> `track.metrics`
`track._log_params` -> `track.params`
`track._log_artifacts` -> `track.artifacts`

File: client/starwhale/api/_impl/track/tracker.py (hunk 2 above)
Comment (tianweidut) [reply client/starwhale/api/_impl/track/tracker.py:335]: we use `self._metric_lock` lock to protect `self._metric_step` and `self._uncommitted_metrics`.

File: client/starwhale/api/_impl/track/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/api/_impl/track/base.py:107]: what i mean is that zerofill is useless when the num is larger than 999, like this :
>>>a=['010','100','200','101','10000', '001']
>>>sorted(a)
output: ['001', '010', '100', '10000', '101', '200']

but the correct sort should: ['001', '010', '100', '101', '200', '10000']

[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/api/_impl/track/base.py]: yes. I have added index max value validation and update key to f`{name}-{index:04}`

**Review by goldenxinxing — APPROVED** [review]
LGTM

**Pull request merged**